### PR TITLE
feat: [#38] Implemented sign in and sign out

### LINKED ISSUE #38: Implement sign in/out

### PR DESCRIPTION
**Pull request by eldarcelik**
### What has changed and why?

- Installed `Formik` package.
- Created a login form with a responsive design and added the `login` route.
- Created schema validation for the login form.
- Created `PrivateRoute` for redirecting unauthorized users.
- Created `login` and `logout` API.
- Created `useAuth` custom hook that will handle sign in/out.
- Used `sessionStorage` to keep auth data.
- Added `logout` icon to the navbar.
- Extended `postData` helper to have request and response types.
- Created types for login.
- Added connection links between login and registration forms.
- Added functionality to show/hide password on forms.
- Added unauthorized info to user before redirecting login page.
- Added `bearer` token in requests headers.
- Refactored `handleResponse` helper to handle responses that do not return content.
- Added `explicit-function-return-type` rule to ESLint
- Closes #38 

### Branch Name Validation

- [x] Branch name pattern: `{type}/{JIRA-ID}` or `{type}/issue-{GITHUB-ISSUE-ID}`

### PR Title Validation

- [x] PR title pattern: `{type}: [{JIRA-ID}] Description` or `{type}: [#{GITHUB-ISSUE-ID}] Description`

### Only check what is relevant for this PR

- [x] Styling verified for different screen sizes and browsers
- [x] Functionality verified for different screen sizes and browsers
- [ ] Relevant unit tests added/updated
- [ ] Storybook files added/updated

### CODE REVIEW

File: src/App.css [hunk 1]
```diff
@@ -14,5 +14,5 @@
 }
 
 .border-danger {
-  border: 1px solid var(--danger-color);
+  border: 1px solid var(--danger-color) !important;
```
Comment (4zuko) [inline src/App.css:17]: Avoid using **`!important`** flag in CSS rules. Try to write more specific CSS rule.

File: src/apis/logout.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+interface ILogout {
+  refreshToken: string;
+}
+
+const logout = async (): Promise<void> => {
+  const refreshToken = sessionStorage.getItem('refreshToken');
+
+  return postData<ILogout, void>(`${process.env.REACT_APP_MOVIES_API_DIRECTUS_URL}/auth/logout`, {
+    refreshToken: refreshToken ? refreshToken : '',
```
Comment (4zuko) [inline src/apis/logout.ts:11]:
```suggestion
    refreshToken: refreshToken || '',
```

File: src/components/Navbar/Navbar.tsx [hunk 3]
```diff
@@ -47,6 +49,7 @@ export default function Navbar() {
         </button>
       </div>
       <input id='search-box' placeholder={`Search for ${searchContent}`} value={search} onChange={onSearchChange} />
+      <i className='fa fa-sign-out' onClick={handleLogout}></i>
```
Comment (4zuko) [inline src/components/Navbar/Navbar.tsx:52]: Accessibility has not been taken into consideration. Please create a new placeholder ticket that could be revisited in the future. Set it for now as a low priority.

File: src/components/Unauthorized/Unauthorized.tsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      navigate('/login', { replace: true });
+    }, DELAY);
+
+    return () => clearTimeout(timeoutId);
+  }, []);
+
+  return (
+    <div className='error'>
+      <h1>Unauthorized Access</h1>
+      <h6>Redirecting to login page...</h6>
```
Comment (4zuko) [inline src/components/Unauthorized/Unauthorized.tsx:21]: Investigate whether using h6 tag is valid from the accessibility point of view. I think that it should be h2. You can always change the font size / any CSS rule.

File: src/helpers/handleResponse.ts [hunk 5]
```diff
@@ -2,11 +2,13 @@ import { STATUS_CODES } from '@/constants/constantValues';
...
 
-const handleResponse = async <T>(response: Response): Promise<T> => {
+const handleResponse = async <R>(response: Response): Promise<R> => {
   if (response.status === STATUS_CODES.OK) {
     const jsonResponse = await response.json();
 
-    return convertKeysToCamelCase<T>(jsonResponse);
+    return convertKeysToCamelCase<R>(jsonResponse);
+  } else if (response.status === STATUS_CODES.NO_CONTENT) {
+    return null as R;
```
Comment (4zuko) [inline src/helpers/handleResponse.ts:11]: If you're retrieving null then, `Promise<R>` should be replaced by `Promise<Nullable<R>>` where `Nullable<T> = T | null`.

File: src/hooks/useAuth.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  const navigate = useNavigate();
+  const [loginState, setLoginState] = useState<ILoginState>(() => {
+    const accessToken = sessionStorage.getItem('accessToken') || null;
+    const refreshToken = sessionStorage.getItem('refreshToken') || null;
+    const expires = parseInt(sessionStorage.getItem('expires') || '0', 10) || null;
+
+    return { accessToken, expires, refreshToken };
+  });
+
+  const handleLoginResponse = (response: ILogin) => {
```
Comment (4zuko) [inline src/hooks/useAuth.tsx:18]: TS return type is missing. Please make sure that you update ESLint rules, so explicitly adding return types is required.

File: src/pages/Account/Login/Login.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+import { ILoginInfo } from './types';
+import { IUser } from '../types';
+
+export default function Login() {
+  const { handleLoginResponse } = useLogin();
+  const [loginInfo, setLoginInfo] = useState<ILoginInfo>({});
+  const [showPassword, setShowPassword] = useState<boolean>(false);
+
+  const handleLogin = (user: IUser, setSubmitting: { (isSubmitting: boolean): void }) => {
+    login(user)
```
Comment (4zuko) [inline src/pages/Account/Login/Login.tsx:20]: Investigate whether promises can be replaced with async/await.

File: src/pages/Account/Login/Login.tsx [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  };
+
+  return (
+    <div className='account-container'>
+      <Formik
+        initialValues={{ email: '', password: '' }}
+        validationSchema={loginSchema}
+        onSubmit={(values, { setSubmitting }) => {
+          handleLogin(values, setSubmitting);
+        }}
```
Comment (4zuko) [inline src/pages/Account/Login/Login.tsx:41]: Try to keep handlers outside of return statement.

**Review by 4zuko — APPROVED** [review]
**`LGTM`**

Nice work. Please check the comments. If some of them will require more time to fix, then create a new Jira ticket for that. Thanks.

File: src/pages/Account/Login/Login.tsx (hunk 7 above)
Comment (eldarcelik) [reply src/pages/Account/Login/Login.tsx:20]: I will create a new issue for this.

File: src/components/Navbar/Navbar.tsx (hunk 3 above)
Comment (eldarcelik) [reply src/components/Navbar/Navbar.tsx]: Please, can you check is it good now?

File: src/apis/getItems.ts [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import { QUERY_TYPE } from '@/constants/constantValues';
 import { getData } from '@/helpers';
-import { ContentType, IItems, QueryType } from '@/types';
+import { IItemsResponse } from '@/pages/Home/types';
```
[the remark this answers is not in the document]
Comment (eldarcelik) [reply src/apis/getItems.ts:3]: I made these changes in all files, but I have some problems with pushing changes from .eslintrc.json file. 
We can look it together at our call.

File: src/helpers/handleResponse.ts (hunk 5 above)
Comment (eldarcelik) [reply src/helpers/handleResponse.ts:11]: `handleResponse` will be deleted in the [next PR](https://github.com/zuko4celik/movies-app-new/pull/45) caused by integrating Axios and interceptors. But, good comment to remember ;)

File: src/App.css (hunk 1 above)
Comment (4zuko) [reply src/App.css:17]: I can still see **`!important`**.

File: src/App.tsx [hunk 10]
```diff
@@ -3,22 +3,39 @@ import React from 'react';
...
+import Registration from './pages/Account/Registration';
+import Error from './pages/Error/Error';
 import Home from './pages/Home';
 import MovieOrShow from './pages/MovieOrShow';
-import Registration from './pages/Registration';
 
-function App() {
+function App(): JSX.Element {
   return (
     <>
```
Comment (4zuko) [inline src/App.tsx:16]: Is there any need for a Fragment here?

File: src/components/Loading/Loading.tsx [hunk 11]
```diff
@@ -4,7 +4,7 @@ import LoadingGif from '@/assets/loading.gif';
 
 import './Loading.css';
 
-export default function Loading() {
+export default function Loading(): JSX.Element {
```
Comment (4zuko) [inline src/components/Loading/Loading.tsx:7]: Investigate what is the difference between **`React.FC<Props>`** and **`JSX.Element`**.

File: src/components/Navbar/Navbar.css [hunk 12]
```diff
@@ -69,6 +70,12 @@
   width: 30%;
 }
 
+#logout-btn {
+  background-color: var(--primary-color);
+  color: var(--secondary-color);
+  cursor: pointer;
+}
+
 @media screen and (max-width: 800px) {
```
Comment (4zuko) [inline src/components/Navbar/Navbar.css:79]: Read more about responsive design and common breakpoints. We should avoid using non-standard values. Please check this https://blog.logrocket.com/css-breakpoints-responsive-design. Thanks.

File: src/App.css (hunk 1 above)
Comment (eldarcelik) [reply src/App.css:17]: Sorry, I missed that.. I fixed it now.

File: src/App.tsx (hunk 10 above)
Comment (eldarcelik) [reply src/App.tsx:16]: No, I am going to delete it.

**Pull request merged**